### PR TITLE
[SYCL][NFC] Small refactoring

### PR DESCRIPTION
**Pull request by romanovvlad**
1. Use C++ cast instead of C casts
2. Move implemenetation of several handler methods to source dir

### CODE REVIEW

File: sycl/include/CL/sycl/handler.hpp [hunk 1]
```diff
@@ -250,7 +249,7 @@ class __SYCL_EXPORT handler {
                             typename std::remove_reference<T>::type>::type>
   F *storePlainArg(T &&Arg) {
     MArgsStorage.emplace_back(sizeof(T));
-    F *Storage = (F *)MArgsStorage.back().data();
+    F *Storage = reinterpret_cast<F *>(MArgsStorage.back().data());
```
Comment (keryell) [inline sycl/include/CL/sycl/handler.hpp:252]: `auto storage`?

File: sycl/include/CL/sycl/handler.hpp [hunk 2]
```diff
@@ -387,7 +374,7 @@ class __SYCL_EXPORT handler {
   }
 
   template <typename T> void setArgHelper(int ArgIndex, T &&Arg) {
-    void *StoredArg = (void *)storePlainArg(Arg);
+    void *StoredArg = static_cast<void *>(storePlainArg(Arg));
```
Comment (keryell) [inline sycl/include/CL/sycl/handler.hpp:377]: `auto StoredArg`

File: sycl/include/CL/sycl/handler.hpp [hunk 3]
```diff
@@ -399,7 +386,7 @@ class __SYCL_EXPORT handler {
   }
 
   void setArgHelper(int ArgIndex, sampler &&Arg) {
-    void *StoredArg = (void *)storePlainArg(Arg);
+    void *StoredArg = static_cast<void *>(storePlainArg(Arg));
```
Comment (keryell) [inline sycl/include/CL/sycl/handler.hpp:389]: auto

File: sycl/include/CL/sycl/handler.hpp [hunk 4]
```diff
@@ -1700,12 +1687,12 @@ class __SYCL_EXPORT handler {
...
       detail::AccessorImplPtr AccImpl = detail::getSyclObjImpl(*AccBase);
 
-      MDstPtr = (void *)AccImpl.get();
+      MDstPtr = static_cast<void *>(AccImpl.get());
       MRequirements.push_back(AccImpl.get());
       MAccStorage.push_back(std::move(AccImpl));
 
       MPattern.resize(sizeof(T));
-      T *PatternPtr = (T *)MPattern.data();
+      T *PatternPtr = reinterpret_cast<T *>(MPattern.data());
```
Comment (keryell) [inline sycl/include/CL/sycl/handler.hpp:1695]: auto

File: sycl/source/detail/scheduler/graph_builder.cpp [hunk 5]
```diff
@@ -455,7 +455,7 @@ Command *Scheduler::GraphBuilder::addHostAccessor(Requirement *Req) {
 Command *Scheduler::GraphBuilder::addCGUpdateHost(
     std::unique_ptr<detail::CG> CommandGroup, QueueImplPtr HostQueue) {
 
-  CGUpdateHost *UpdateHost = (CGUpdateHost *)CommandGroup.get();
+  CGUpdateHost *UpdateHost = static_cast<CGUpdateHost *>(CommandGroup.get());
```
Comment (keryell) [inline sycl/source/detail/scheduler/graph_builder.cpp:458]: auto

File: sycl/CMakeLists.txt [hunk 6]
```diff
@@ -14,7 +14,7 @@ include(AddSYCLExecutable)
 set(SYCL_MAJOR_VERSION 2)
 set(SYCL_MINOR_VERSION 1)
 set(SYCL_PATCH_VERSION 0)
-set(SYCL_DEV_ABI_VERSION 3)
+set(SYCL_DEV_ABI_VERSION 4)
```
Comment (alexbatashev) [inline sycl/CMakeLists.txt:17]: It's not a breaking change. Other lines just got sorted in another way.

File: sycl/include/CL/sycl/handler.hpp [hunk 7]
```diff
@@ -308,8 +307,8 @@ class __SYCL_EXPORT handler {
   /// Streams are then forwarded to command group and flushed in the scheduler.
   ///
   /// \param Stream is a pointer to SYCL stream.
-  void addStream(shared_ptr_class<detail::stream_impl> Stream) {
-    MStreamStorage.push_back(std::move(Stream));
+  void addStream(const shared_ptr_class<detail::stream_impl> &Stream) {
+    MStreamStorage.push_back(Stream);
```
Comment (v-klochkov) [inline sycl/include/CL/sycl/handler.hpp:311]: Not about your change: When I implemented handler::addReduction() adding objects to MSharedPtrStorage I thought that MStreamStorage could be deleted and MSHaredPtrStorage could be used instead. That would save the field without any bad consequences.

Comment (romanovvlad) [reply sycl/include/CL/sycl/handler.hpp:311]: Hm, it seems you are right. Will do.
I think refactoring patches are good candidates for "not related" comments.

Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:311]: Understood, thank you.

**Comment by v-klochkov** [comment]
Looks good, but accessor.cpp LIT test reported a fail and it is known as a stable test.

**Comment by romanovvlad** [comment]
> 
> 
> Looks good, but accessor.cpp LIT test reported a fail and it is known as a stable test.

It's not old one accessor from basic_tests, this one from atomic_ref and has been added quite recently.
I have seen this fail in another PR, so it's not related to the change.
Trying to root cause the fail, but it turns to be difficult one.

**Comment by romanovvlad** [comment]
> 
> 
> > Looks good, but accessor.cpp LIT test reported a fail and it is known as a stable test.
> 
> It's not old one accessor from basic_tests, this one from atomic_ref and has been added quite recently.
> I have seen this fail in another PR, so it's not related to the change.
> Trying to root cause the fail, but it turns to be difficult one.

atomic_ref/accessor should start passing after recent update of L0.

**Pull request merged**